app/gui/view/graph-editor/src/lib.rs

```diff
@@ -3652,3 +3649,170 @@ impl display::Object for GraphEditor {
+        let (_, ref graph_editor) = init();
+        let edges = graph_editor.edges();
+        assert!(graph_editor.nodes().is_empty());
+        assert!(edges.is_empty());
+        // Adding two nodes.
+        let (node_id_1, node_1) = graph_editor.add_node_by_api();
+        graph_editor.stop_editing();
+        let (node_id_2, node_2) = graph_editor.add_node_by_api();
+        graph_editor.stop_editing();
+        // Creating edge.
+        let port = node_1.model.output_port().expect("No output port");
+        port.events.mouse_down.emit(());
+        let edge_id = graph_editor.on_edge_add.value();
+        let edge = edges.get_cloned_ref(&edge_id).expect("Edge was not added");
+        assert_eq!(edge.source().map(|e| e.node_id), Some(node_id_1));
+        assert_eq!(edge.target().map(|e| e.node_id), None);
+        assert_eq!(edges.len(), 1);
+        // Connecting edge.
+        // We need to enable ports. Normally it is done by hovering the node.
+        node_2.model.input.frp.set_ports_active(true, None);
+        let port = node_2.model.input_port().expect("No input port");
+        port.hover.events.mouse_down.emit(());
+        assert_eq!(edge.source().map(|e| e.node_id), Some(node_id_1));
+        assert_eq!(edge.target().map(|e| e.node_id), Some(node_id_2));
+    }
+
+
+    // === Test utilities ===
+
+    struct Case {
+        node_source: Option<NodeId>,
+        should_edit: bool,
+    }
+
+    impl GraphEditor {
+        fn add_node_by<F: Fn(&GraphEditor)>(&self, add_node: &F) -> (NodeId, Node) {
+            add_node(self);
+            let (node_id, ..) = self.node_added.value();
+            let node = self.model.nodes.all.get_cloned_ref(&node_id).expect("Node was not added");
+            node.set_expression(node::Expression::new_plain("some_not_empty_expression"));
+            (node_id, node)
+        }
+
+        fn add_node_by_api(&self) -> (NodeId, Node) {
+            let add_node = |editor: &GraphEditor| editor.add_node();
+            self.add_node_by(&add_node)
+        }
+
+        fn assert(&self, case: Case) {
+            let (added_node, node_source, should_edit) = self.node_added.value();
+            let node_being_edited = self.node_being_edited.value();
+            assert_eq!(
+                should_edit, case.should_edit,
+                "Node editing state does not match expected."
+            );
+            assert_eq!(should_edit, node_being_edited.is_some());
+            if let Some(node_being_edited) = node_being_edited {
+                assert_eq!(node_being_edited, added_node, "Edited node does not match added one.");
+            }
+            let node_source = node_source.map(|source| source.node);
+            assert_eq!(node_source, case.node_source, "Source node does not match expected.");
+        }
+
+        fn nodes(&self) -> &Nodes {
+            &self.model.nodes
+        }
+
+        fn edges(&self) -> &Edges {
+            &self.model.edges
+        }
+    }
```

Comment:
Let's make it methods in graph-editor's lib.rs

Reply:
Done, also replaced all found usages of such pattern.